File: leetcode-everyday/搜索算法/40-组合总和II/40-组合总和II.cpp
class Solution {
private:
    void dfs(vector<int>& candidates, int target, int start, vector<bool>& visited, vector<int>& path, vector<vector<int>>& ans) {
        if(target == 0) {
            ans.push_back(path);
            return;
        }
        for(int i = start; i < candidates.size(); ++i) {
            if(visited[i]) continue;
            if(i > 0 && candidates[i] == candidates[i-1] && !visited[i-1]) continue;
            if(target-candidates[i] < 0) break;
            visited[i] = true;
            path.push_back(candidates[i]);
            dfs(candidates, target-candidates[i], i+1, visited, path, ans);
            visited[i] = false;
            path.pop_back();
        }
    }
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        vector<vector<int>> ans;
        int m = candidates.size();
        vector<bool> visited(m, false);
        vector<int> path;
        sort(candidates.begin(), candidates.end());
        dfs(candidates, target, 0, visited, path, ans);
        return ans;
    }
};

File: leetcode-everyday/搜索算法/40-组合总和II/40-组合总和II.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
 var combinationSum2 = function(candidates, target) {
  let m = candidates.length;
  candidates.sort((a, b) => a-b);
  const visited = new Array(m).fill(false);
  const ans = [];
  dfs(candidates, target, 0, 0, [], visited, ans);
  return ans;
};

function dfs(candidates, target, start, pos, path, visited, ans) {    
  if(pos === target) {
      ans.push([...path]);
      return;
  }
  for(let i = start; i < candidates.length; ++i) {
      if(visited[i]) continue;
      if(i > 0 && candidates[i] === candidates[i-1] && !visited[i-1]) continue;
      if(pos + candidates[i] > target) break;
      visited[i] = true;        
      dfs(candidates, target, i+1, pos + candidates[i], [...path, candidates[i]], visited, ans);
      visited[i] = false;
  }
}
